UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,548 @@
+          StartAddress  = Fdt64ToCpu (*Data64);
+          NumberOfBytes = Fdt64ToCpu (*(Data64 + 1));
+          // DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          // DEBUG ((DEBUG_INFO, "  %016lX  %016lX", StartAddress, NumberOfBytes));
+        } else if (AsciiStrCmp (TempStr, "ecc-detection-bits") == 0) {
+          Data32  = (UINT32 *)(PropertyPtr->Data);
+          ECCData = Fdt32ToCpu (*Data32);
+        } else if (AsciiStrCmp (TempStr, "ecc-correction-bits") == 0) {
+          Data32   = (UINT32 *)(PropertyPtr->Data);
+          ECCData2 = Fdt32ToCpu (*Data32);
+          // DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          // DEBUG ((DEBUG_INFO, "  %X", Attribute));
+        }
+      }
+
+      if (ECCData == ECCData2) {
+        if (ECCData == 1) {
+          ECCAttribute = EFI_RESOURCE_ATTRIBUTE_SINGLE_BIT_ECC;
+        } else if (ECCData == 2) {
+          ECCAttribute = EFI_RESOURCE_ATTRIBUTE_MULTIPLE_BIT_ECC;
+        }
+      }
+
+      if (ECCAttribute != 0) {
+        Attribute |= ECCAttribute;
+      }
+
+      BuildResourceDescriptorHob (EFI_RESOURCE_SYSTEM_MEMORY, Attribute, StartAddress, NumberOfBytes);
+
+      /*if (AsciiStrCmp (TempStr, "memory") == 0) {
+        BuildResourceDescriptorHob (EFI_RESOURCE_SYSTEM_MEMORY, Attribute, StartAddress, NumberOfBytes);
+      }
+      else if (AsciiStrCmp (TempStr, "mmio") == 0) {
+        BuildResourceDescriptorHob (EFI_RESOURCE_MEMORY_MAPPED_IO, Attribute, StartAddress, NumberOfBytes);
+      }
+      else if (AsciiStrCmp (TempStr, "reserved") == 0) {
+        BuildResourceDescriptorHob (EFI_RESOURCE_MEMORY_RESERVED, Attribute, StartAddress, NumberOfBytes);
+      }*/
+    } // end of memory node
+    // reserved-memory
+    else if (AsciiStrCmp (NodePtr->Name, "reserved-memory") == 0) {
+      for (SubNode = FdtFirstSubnode (Fdt, Node); SubNode >= 0; SubNode = FdtNextSubnode (Fdt, SubNode)) {
+        NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + SubNode + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+        DEBUG ((DEBUG_INFO, "\n      SubNode(%08X)  %a", SubNode, NodePtr->Name));
+
+        PropertyPtr = FdtGetProperty (Fdt, SubNode, "reg", &TempLen);
+        ASSERT (TempLen > 0);
+        if (TempLen > 0) {
+          Data64        = (UINT64 *)(PropertyPtr->Data);
+          StartAddress  = Fdt64ToCpu (*Data64);
+          NumberOfBytes = Fdt64ToCpu (*(Data64 + 1));
+          // DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          // DEBUG ((DEBUG_INFO, "  %016lX  %016lX", StartAddress, NumberOfBytes));
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "mmio@", AsciiStrLen ("mmio@")) == 0) {
+          DEBUG ((DEBUG_INFO, "  MemoryMappedIO"));
+          BuildMemoryAllocationHob (StartAddress, NumberOfBytes, EfiMemoryMappedIO);
+        } else {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "compatible", &TempLen);

```

Comment:
Are we going to make compatible mandatory?

Reply:
I also see stuff like #if 0...Is this patch cleaned up and ready for merge?

Reply:
**Are we going to make compatible mandatory?**
It based on 
https://docs.google.com/document/d/1WxEUlCsXpc17DkJhL3XVkOW7e_KIt_zcc9tQ1trOySg/edit#heading=h.sryh1uptwsvb 7.8 reserved-memory (R)

### I also see stuff like #if 0...Is this patch cleaned up and ready for merge?
Will remove the  directives, controls compilation of portions of a source file before merging.

Reply:
Noted. But without the compatible string , we don't have any clue about the attribute of reserved memory nodes so far.
just as you mentioned.

Reply:
EfiReservedMemoryType Can't we use this one?

Reply:
Ok , will try that and update the result.
Thanks